ENH Fix unfriendly error message for documentation checks

chibby0ne opened the pull request:
Fixes: https://github.com/scikit-learn/scikit-learn/issues/11673


Previously error messages for mismatches parameters on function/method
signatures and docstrings were not explicit or user-friendly. This
changes provides a more explicit error message and in addition notifies
the relation of missing/mismatching parameters from signature to
docstring and from docstring to signature.


Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

chibby0ne committed:
utils/testing: improve error message for doc check

Previously error messages for mismatches parameters on function/method
signatures and docstrings were not explicit or user-friendly. This
changes provides a more explicit error message and in addition notifies
the relation of missing/mismatching parameters from signature to
docstring and from docstring to signature.

Fixes: https://github.com/scikit-learn/scikit-learn/issues/11673

Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

jnothman reviewed (commented):
Why not just show a standard diff output with `>` and `<`...? Order matters here, not just the set of names, doesn't it?

chibby0ne commented:
Hi @jnothman! Thank you so much for your input. 

> Why not just show a standard diff output with > and <...? 

If I understood you correctly, you would prefer to have a kind of diff output between docstrings and function signature? 
If that's the case, I looked into this yesterday and while it would be possible to clearly show in the docstring where something might be missing or mistaken, the same cannot be said of the function signature i.e: if there are parameters only in the signature since the signature are normally written in one line they wouldn't look very good using the diff `>` or `<` way. 

> Order matters here, not just the set of names, doesn't it?

That's correct. That is taken care on the `else` part of the statement, which handles the cases where the number of parameters in both the docstrings and the function signature match.

At the moment, If you get the order wrong but have the same number of parameters in both the docstring and the function signature then you would receive this kind of error notifications:

```
>           raise AssertionError("Docstring Error: " + msg)                                                                                                                                                                                                                     
E           AssertionError: Docstring Error:                                                                                                                                                                                                                                    
E           sklearn.linear_model.coordinate_descent.ElasticNetCV.__init__ positive != random_state                                                                                                                                                                              
E           sklearn.linear_model.coordinate_descent.ElasticNetCV.__init__ random_state != positive                                                                                                                                                                              
```

In this particular case I switched the order of these two are adjacent parameters in the signature. 

A case could be made that this error message is also not very helpful. I would propose having an error message like this (for the same previous example):

```
>           raise AssertionError("Docstring Error: " + msg)                                                                                                                                                                                                                     
E           AssertionError: Docstring Error:                                                                                                                                                                                                                                    
E           There's a difference in the order of the parameters in sklearn.linear_model.coordinate_descent.ElasticNetCV.__init__'s signature and its docstring.                                                                                                                 
E           According to the function signature the parameter in this place of the docstring should be: "positive" but instead is "random_state"                                                                                                                                
E           There's a difference in the order of the parameters in sklearn.linear_model.coordinate_descent.ElasticNetCV.__init__'s signature and its docstring.                                                                                                                 
E           According to the function signature the parameter in this place of the docstring should be: "random_state" but instead is "positive"
```

chibby0ne committed:
utils/testing: adapt testing.py UTs for new changes

- Adjust lines to less than 80 chars.
- Adapt all unit tests checking docstring's test util function.

Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

jnothman commented:
Actually, I think pytest does a pretty good job of comparing lists, you just need to use that.

Running pytest on:

```py
    def test_blah():
        assert ['a', 'b', 'c'] == ['a', 'd', 'b'], "foobar"
```

Gives the following error. See especially the diff.

```pytb
    def test_blah():
>       assert ['a', 'b', 'c'] == ['a', 'd', 'b'], "foobar"
E       AssertionError: foobar
E       assert ['a', 'b', 'c'] == ['a', 'd', 'b']
E         At index 1 diff: 'b' != 'd'
E         Full diff:
E         - ['a', 'b', 'c']
E         + ['a', 'd', 'b']

```

chibby0ne committed:
testing: Improve docstring error reporting

Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

chibby0ne commented:
Hi @jnothman thank you for your feedback! I'm sorry I took so long to get back into this PR. I've made the changes according to your suggestions. I've looked into how pytest creates this and added the same functionality. The errors now would look something like this:

```
Parameters in function docstring have more items, first extra item: c"                                                 
Full diff:                 
- ['a', 'b']           
+ ['a', 'b', 'c']"
?          +++++
```

chibby0ne commented:
Let's not mege just yet... I found discrepancies on how the error is printed in the tests, and how they show up when an actual docstring error is found. I'll take a look into it today.

chibby0ne committed:
Fix formatting of docstrings assertion error message

Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

chibby0ne commented:
Ok so now I've made a tiny fix into how this was being shown, and I've finally verified that it works correctly with actual docstring error by removing a parameter in the docstring of two functions with different number of parameters, which throws the following error:

```
>           raise AssertionError("Docstring Error:\n" + msg)                                                       
E           AssertionError: Docstring Error:
E           In :<function ElasticNetCV.__init__ at 0x7f1e1955e048>                                                 
E           At index 13 diff: 'positive' != 'random_state'                                                         
E           Parameters in function signature have more items, first extra item: selection                          
E           Full diff:
E           ['l1_ratio',
E           'eps',
E           'n_alphas',
E           'alphas',
E           'fit_intercept',
E           'normalize',
E           'precompute',
E           'max_iter',
E           'tol',
E           'cv',
E           'copy_X',
E           'verbose',
E           'n_jobs',
E           -  'positive',
E           'random_state',
E           'selection']
E           In :<function BayesianRidge.fit at 0x7f1e196648c8>                                                     
E           At index 1 diff: 'y' != 'sample_weight'
E           Parameters in function signature have more items, first extra item: sample_weight                      
E           Full diff:
E           - ['X', 'y', 'sample_weight']
E           ?      -----
E           + ['X', 'sample_weight']
```

@jnothman what do you think? Any suggestions?

jnothman commented:
Thanks for that, it looks like a great improvement

jnothman commented on sklearn/utils/testing.py:
```diff
@@ -953,11 +953,13 @@ def check_docstring_parameters(func, doc=None, ignore=None, class_name=None):
+    if len(param_signature) > 0 and param_signature[0] == 'self':
+        param_signature.remove('self')
 
+    # Analize function's docstring
```
*Analyze

jnothman reviewed (approved):
Nice work! My concern about the code is to make sure you were licensed to contribute it here: is it your work, or is it under a compatible licence (and probably in need of attribution)?

jnothman commented on sklearn/utils/testing.py:
```diff
@@ -980,16 +983,55 @@ def check_docstring_parameters(func, doc=None, ignore=None, class_name=None):
+    incorrect.extend(message)
+
+    # Prepend function name
+    incorrect = ['In :' + str(func)] + incorrect
```
The space before : looks strange especially if there is no space after it.

chibby0ne commented:
@jnothman mostly taken from pytest, but it is licensed with a [compatible license](https://github.com/pytest-dev/pytest/blob/master/LICENSE) i.e: MIT. You're right we should probably attribute it to it.
How do you suggest I do this? Just add stating: This logic comes from the assertion statement in the pytest project.

I'll implement comments later today. Thanks for your feedback!

chibby0ne committed:
utils/testing: Fix typo in comment and in error

Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

chibby0ne commented:
@jnothman TBH I'm not sure what to do. What do you suggest? 
Let's say we wanted to include the whole MIT license with copyright where should I write it? Top of the file? Close to the actual lines?

jnothman reviewed (commented):
It should be okay to just say "the following is derived from ..., Copyright ..., Licensed under ... See http..."

chibby0ne committed:
utils/testing: Add attribution of code from pytest

Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

chibby0ne commented:
Thanks for the clarification and quick reply @jnothman! I've added a comment as per your suggestion.

chibby0ne committed:
utilts/testing: fix unit tests and tidy error msg

- Removed class parameter from check_docstring_parameters function,
since we get the class name from the attribute __qualname__ in Python 3
and .im_class.__name__ in Python 2.

Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

chibby0ne commented:
Hi @jnothman I've updated/fixed the unit tests for `utils/testing` and improved a bit the error message part specifying the function at which the error was found, previously it would report also the memory address but now it would contains the whole module.class.function name.
So instead of:
```
E           AssertionError: Docstring Error:                                                                                                                                                                                                  
E           In :<function BayesianRidge.fit at 0x7fa92beff840>                                                                                                                                                                                
E           At index 1 diff: 'y' != 'yy'                                                                                                                                                                                                      
E           Full diff:                                                                                                                                                                                                                        
E           - ['X', 'y', 'sample_weight']                                                                                                                                                                                                     
E           + ['X', 'yy', 'sample_weight']                                                                                                                                                                                                    
E           ?        +      
```
It would be:
```
E           AssertionError: Docstring Error:
E           In function: sklearn.linear_model.bayes.BayesianRidge.fit
E           At index 1 diff: 'y' != 'yy'                                                                                                                                                                                                      
E           Full diff:                                                                                                                                                                                                                        
E           - ['X', 'y', 'sample_weight']                                                                                                                                                                                                     
E           + ['X', 'yy', 'sample_weight']                                                                                                                                                                                                    
E           ?        +      
```

chibby0ne commented:
@jnothman Do we need to update the PR title with a `[MRG + 1]`?

chibby0ne commented:
Seems like this PR has fallen off the radar. I guess I need another maintainer to approve it. What should I do to finalize it?  BTW Should I squash the commits?

amueller commented:
This looks great. But is the string in the test the full string that's shown?
I didn't see "contains a parameter mismatch between the function signature and its docstring" but maybe I'm tired :-/

chibby0ne commented:
> This looks great. But is the string in the test the full string that's shown?
> I didn't see "contains a parameter mismatch between the function signature and its docstring" but maybe I'm tired :-/

@amueller  No you're right! That string is no longer the string shown for that kind of error. Also the string shown depends on the type of error:

* Typos (like the ones shown in three comments above this thread):
```
           ["In function: sklearn.utils.tests.test_testing.f_bad_order",
             "At index 0 diff: 'b' != 'a'",
             "Full diff:",
             "- ['b', 'a']",
             "+ ['a', 'b']"],
```

* If # parameter's in docstring > # parameters in signature:
```
            ["In function: " +
                "sklearn.utils.tests.test_testing.f_too_many_param_docstring",
             "Parameters in function docstring have more items, first extra"
             + " item: c",
             "Full diff:",
             "- ['a', 'b']",
             "+ ['a', 'b', 'c']",
             "?          +++++"],
```

* If # parameter's in signature > # parameters in docstring:

```
            ["In function: sklearn.utils.tests.test_testing.f_missing",
             "Parameters in function signature have more items, first extra"
             + " item: b",
             "Full diff:",
             "- ['a', 'b']",
             "+ ['a']"],
```

adrinjalali commented:
@chibby0ne , any chance you would have the time to finish the last bits here any time soon? Looks almost finished.

chibby0ne commented:
@adrinjalali Yep it is almost there. I will make the finishing changes later today.
@amueller Agreed! Kudos for using "uninitiated", read it with Bane's voice in my mind :smile:

chibby0ne committed:
testing: make docstring error message more explicit

Since these are errors regarding the docstring, having a explicit
message indicating it goes a long way into debugging when tests fail due
to it.

Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

chibby0ne commented:
Sorry for the delay guys.
Here are the updated messages:
I've also taken the liberty to rebase the PR since I noticed it had a confict.

* Typos (like the ones shown in three comments above this thread):
```
                ["In function: sklearn.utils.tests.test_testing.f_bad_order",    
                 "There's a parameter name mismatch in function docstring w.r.t."
                 " function signature, at index 0 diff: 'b' != 'a'",                                                                                                                                                                          
                 "Full diff:",                                                                                                                                                                                                                 
                 "- ['b', 'a']",                                                                                                                                                                                                              
                 "+ ['a', 'b']"],  
```

* If # parameter's in docstring > # parameters in signature:
```
                ["In function: " +                                                                                                                                                                                                            
                    "sklearn.utils.tests.test_testing.f_too_many_param_docstring",                                                                                                                                                            
                 "Parameters in function docstring have more items w.r.t. function"                                                                                                                                                           
                 " signature, first extra item: c",                                                                                                                                                                                           
                 "Full diff:",                                                                                                                                                                                                                
                 "- ['a', 'b']",                                                                                                                                                                                                              
                 "+ ['a', 'b', 'c']",                                                                                                                                                                                                         
                 "?          +++++"], 
```

* If # parameter's in signature > # parameters in docstring:

```
                ["In function: sklearn.utils.tests.test_testing.f_missing",                                                                                                                                                                   
                 "Parameters in function docstring have less items w.r.t. function"                                                                                                                                                            
                 " signature, first missing item: b",                                                                                                                                                                                         
                 "Full diff:",                                                                                                                                                                                                                
                 "- ['a', 'b']",                                                                                                                                                                                                               
                 "+ ['a']"],      
```

What do you think  @amueller , @adrinjalali ?

chibby0ne commented:
I just realized this PR hasn't been merged yet. I noticed that the codecov/patch check didn't pass and I didn't quite understand the coverage report tbh. Maybe I'm missing something but it seems like the tests are included in the coverage report.
@jnothman can you give me some suggestions on how to pass this coverage/patch check and get this PR finally merged?

jnothman commented:
Yes, this seems a low risk change that we should merge... But you need someone else's approval at least

rth commented on sklearn/utils/testing.py:
```diff
@@ -904,16 +901,21 @@ def _get_func_name(func, class_name=None):
+        if qualname != func.__name__:
+            parts.append(qualname[:qualname.find('.')])
+    except AttributeError:
+        # Python 2
```
We no longer support python2, so I think this part can be removed.

chibby0ne replied:
Good catch! Will do!

rth reviewed (approved):
Minor comment, apart from that LGTM.

chibby0ne committed:
utils/testing: Remove try/except for Python2 cases

Python2 is no longer supported

Signed-off-by: Antonio Gutierrez <chibby0ne@gmail.com>

rth commented:
Thanks a lot! Merging.

The CircleCI failure is due to an outdated branch; it should be fine on master.

chibby0ne commented:
> Thanks a lot! Merging.
> 
> The CircleCI failure is due to an outdated branch; it should be fine on master.

Great! Thank you for the support @rth!
